---
title: "week9"
author: "rajkomalkaur"
date: "2025-09-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 

```{r}
suppressPackageStartupMessages({
  library("seqinr") # is a package designed to process and analyse sequence data.
  library("R.utils") # general utilities like zip and unzip
})
```

## Loading E.coli seuence data

We will be downloading E. coli genomic sequence data from Ensembl. Ensembl is one of the biggest databases of genomic data

```{r download}
library("R.utils")

URL="http://ftp.ensemblgenomes.org/pub/bacteria/release-53/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"
download.file(URL,destfile="ecoli_cds.fa.gz")
#gunzip("ecoli_cds.fa.gz")
list.files()
```
## Load the library
This library provides tools for biological sequence retrieval and analysis
```{r}
library("seqinr")
cds <- seqinr::read.fasta("ecoli_cds.fa")
str(head(cds))
```
## Basic Sequence analysis
we can figure out the number of coding sequences and generate a detailed summary of the sequences, including their lengths, classes, and modes, while head limits the output to the initial entries.
```{r}
length(cds)
head(summary(cds))
```
## determine the total length of our sequence.
he code below extracts the lengths of genes from the cds object by converting the first column of the summary to numeric values. It then calculates the total length of all genes by summing these values.
```{r}
len <- as.numeric(summary(cds)[,1])

sum(len)
mean(len)
median(len)
hist(len)
```
#plot histogram
```{r}
hist(len,xlab="sequence length (bp)",ylab="frequency",main="histogram of coding sequence length")
boxplot(len,ylab="sequence length (bp)")
```
## Sequence Composition
CDS contains multiple sequences.o access a specific element within a list, such as the first sequence, we use double square brackets ([[ ]]). This syntax retrieves the actual sequence object from the list. For example, cds[[1]] extracts the first sequence from the cds list. We can then pass this to the GC function to compute its GC content.
```{r}
GC(cds[[1]])
count(cds[[1]],1)
count(cds[[1]],2)
count(cds[[1]],3)
```
```{r}
length(unlist(cds[1:3]))
GC(unlist(cds[1:3]))
```
## Analysed nucleotide composition
 to analyze the nucleotide composition of a DNA sequence, we first need to unlist the cds, store it in the dna variable and then calculate the frequency of each nucleotide using dna_composition <- count(dna, 1), which stores these frequencies in dna_composition. We then visualize this data with barplot(dna_composition, xlab="nucleotides", ylab="frequency", main="E coli CDS composition"). This creates a bar plot where the x-axis represents the nucleotides (A, C, G, T), the y-axis shows their frequencies, and the plot title indicates that it displays the nucleotide composition of E. coli coding sequences.
```{r}
dna <- unlist(cds)
dna_composition <- count(dna,1)

barplot(dna_composition,xlab="nucleotides",ylab="frequency", main="E coli CDS composition")
```
# find relative frequencies
To find the relative frequencies, we divide each nucleotide count by the total number of nucleotides using dna_composition / sum(dna_composition). This step transforms the raw counts into proportions, showing the frequency of each nucleotide as a fraction of the total, which helps in understanding the distribution of nucleotides within the sequence.
```{r}
dna_proportion <- dna_composition/sum(dna_composition)
```
# barplot
```{r}
barplot(dna_proportion,xlab="nucleotides",ylab="proportion", main="E coli CDS composition")

grid()
```
# protein sequence analysis
```{r}
translate(cds[[1]])
```
#loop for translation
In the code, lapply applies the translate function to each element of the cds list, resulting in prot, a list of translated protein sequences.
```{r}
prot <- lapply(cds, translate)

aa <- unique(prot[[2]])
aa <- aa[aa != "*"]
length(aa)
```
#determine frequency of amino acids
determine the frequency of each amino acid in our first gene. This code counts the occurrences of each amino acid in the first sequence of the prot list, using aa as the set of possible amino acids. 
```{r}
count(prot[[1]],wordsize=1,alphabet=aa)
count(prot[[2]],wordsize=1,alphabet=aa)
```
# Codon usage
A codon is a three base triplet that codes for a particular amino acid. As there are 64 possible codons and just 20 amino acids (and 3 termination codons), some amino acids are coded by multiple codons. For example, leucine is coded by six codons.
```{r}
uco(cds[[2]])
```



```{r}
uco(cds[[2]],index="rscu")
```

```{r}
uco(cds[[2]],index="rscu",as.data.frame=TRUE)
```
#kmer profiling
K-mer is simply a short sequence string of length k. We can use k-mer analysis to identify particular sequences that are over- or under-represented in an organism.
```{r}
prots <- unlist(prot)

mycounts <- count(prots,wordsize=3,alphabet=aa)

str(mycounts)
```
```{r}
head(mycounts)
myfreq <- count(prots,wordsize=3,alphabet=aa,freq=TRUE)
```
